03-task-manager/optionalArrayMethodsReviewExtraAssignment.js

```diff
@@ -0,0 +1,291 @@
+ */
+const transformWordIntoTitle = (characterInWord, indexOfCharacter) => {
+  // We only want to change the FIRST letter of the word to uppercase
+  if (indexOfCharacter === 0) {
+    return characterInWord.toUpperCase();
+  } // we have returned!! The rest of the code will ONLY run for characters
+    // after the first one
+
+  // We could skip `.toLowerCase`, but if a letter in the middle of the word
+  // is uPpErcAse, it'll look nicer if we transform it into lowercase
+  return characterInWord.toLowerCase();
+}
+
+/**
+ * This is the callback used when we are mapping over an array of "words," like:
+ *
+ * ```
+ * ["Carlos", "d.", "Perez"]
+ * ```
+ *
+ * This function receives a string (just ONE of those words, like "d.").
+ *
+ * It will split the word up into an array of letters, use the map function
+ * from before to transform that array into title-case, then join that
+ * transformed array back into a string, and return the result.
+ *
+ * This is the most wasteful & superfluous step. You probably notice we could
+ * just do this instead:
+ *
+ * ```
+ * firstLetter = wordInString[0];
+ * otherLetters = wordInString.splice(1);
+ * return `${firstLetter.toLowerCase()}${otherLetters.toLowerCase()}`
+ * ```
+ *
+ * Indeed, this would be much faster since we avoid an inner loop, but our goal
+ * is to learn, not to go fast!
+ */
+const transformStringIntoTitledWords = (wordInString) => {
+  const letters = wordInString.split('');
+  const titleCaseLetters = letters.map(transformWordIntoTitle);
+  return titleCaseLetters.join('');
+}
+
+/**
+ * Finally, the highest level: this callback operates on every string in our
+ * main name array for this example. It breaks the name up into an array of
+ * words first:
+ *
+ * ```
+ * "carlos cantiago"
+ *    -> ["carlos", "cantiago"]
+ *      -> ["Carlos", "Santiago"]
+ *        -> "Carlos Santiago"
+ * ```
+ */
+const transformNameIntoTitleCase = (name) => {
+  // We'll use a regex to split the string. ' +' means "one or more spaces."
+  // This is good because it'll work for our name "tam  person" where there is
+  // a double-space
+  const nameWords = name.split(/ +/);
+  const titleCaseWords = nameWords.map(transformStringIntoTitledWords)
+  return titleCaseWords.join(' ');
+}
+
+console.log(
+  'titledNames verbose',
+  names.map(transformNameIntoTitleCase)
+)
+
+
+//////// CHALLENGE: Remove names with the wrong format
+//                  AND change it to "Title Case"
+//                  AND remove people whose last name ends with z
+//                  AND write a message asking them to sign up
+const result = names
+  // remove bad format
+  .filter((name) => name.match(rightFormat))
+  // change to title case
+  .map(transformNameIntoTitleCase)
+  // remove names that end in "z"
+  .filter((name) => {
+    const lastLetter = name.slice(-1);
+    return lastLetter.toLowerCase() !== 'z'
+  })
+  // transform into a sign-up message
+  .map((name) => `
+    <h1>Hey there ${name}!</h1>
```

Comment:
Maybe explain why you are emitting HTML?

Reply:
There's not really any reason, so I just changed it to plain text